README.md
## 'Go to Matching Bracket' in Brackets editor

[![Semantic Versioning](https://img.shields.io/github/release/davidwaterston/goto-matching-bracket.svg)](http://semver.org/)
[![MIT Licence](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/davidwaterston/goto-matching-bracket/blob/master/LICENSE)
[![Releases signed with Gnu Privacy Guard](https://img.shields.io/badge/gpg-signed-green.svg)](#verifying-releases)
[![Stories in Progress](https://badge.waffle.io/davidwaterston/goto-matching-bracket.svg?label=in%20progress&title=in progress)](http://waffle.io/davidwaterston/goto-matching-bracket)
[![Join the Chat](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/davidwaterston/goto-matching-bracket)


A [Brackets](http://brackets.io/) extension to instantly locate and place the cursor on the matching bracket to the one under the cursor. Adds an entry called _Go to Matching Bracket_ to the _Navigate_ menu with a shortcut of Command-Alt-&#8594; (Mac) or Ctrl-Alt-&#8594; (Windows). **Note that this shortcut was changed in version 2.0.0 as release 1.3 of Brackets now uses the original shortcut for _Expand Current_**.

The 'brackets' characters that are matched are (, ), {, }, [ and ].

<img src="https://dvolvr.files.wordpress.com/2015/04/goto-matching-bracket-v2.gif" width="800" alt="Brackets Goto Matching Bracket in action" />


## Install

The best way to install this extension is via the Brackets Extension Manager using the _File > Extension Manager..._ menu item. When the dialog opens you can search for the extension or, alternatively, click on the _Install from URL_ button and enter:
```
https://github.com/davidwaterston/goto-matching-bracket
```
If you prefer to install using Git then you can pull directly from this repo:

Open Brackets and click _Help > Show Extensions Folder_.
Clone the extension into the user subdirectory:

```
git clone git://github.com/davidwaterston/goto-matching-bracket.git user/davidwaterston.goto-matching-bracket
```


##Languages
'Go to Matching Bracket' is currently available in:
 - Czech
 - English
 - French
 - German
 - Italian
 - Polish
 - Portuguese
 - Russian

If your preferred language is not included the extension will still work but the (single) menu item will appear in English.
Should you wish to see it translated into your own language (or you think one of the existing translations is wrong) please let me know by commenting in [_Issues_](https://github.com/davidwaterston/goto-matching-bracket/issues).


## Preferences
There are two preferences which can be set in your Brackets Preferences File to override the default behaviour of this extension:

| Preference | Description | Default Value |
| :--------- | :---------- | :------------ |
| maxScanLines | The maximum number of lines that will be scanned (searched) to find a match. If, after scanning this number of lines, no match is found then the search will give up _even if a matching bracket exists in the file_. Unless you are working on very large files you are unlikely to want to change this value. | 5000 |
| maxScanLineLength | The maximum length that a line may be to be considered for scanning. Lines that are longer than this length will be skipped (the scan will continue on to the next line, if we have not reached the start/end of the file). For example, if this value is set to 10 then any line whose length is greater than 10 will be ignored when looking for a matching bracket. This, of course, means that it would be possible to miss a match if this value is set too low. | 10000 |

Should you find that you need to override any of the default values this can be done by editing your preference file using the Brackets _Debug > Open Preferences File_ menu item and adding one, or more, of the following to it:

```
"com.github.davidwaterston.gotoMatchingBracket.maxScanLines": 20000,
"com.github.davidwaterston.gotoMatchingBracket.maxScanLineLength": 1000,
```

You should, of course, alter the values (after the ':') to the required value.


## Compatibility
This extension has been tested and confirmed to work on Brackets versions 1.2 to 1.5 inclusive.


## Frequently Asked Questions
**Q: Why doesn't it do anything?**  
A: There are a few reasons the extension might not appear to be working. If you are working on a very large file (> 5000 lines) and the opening and closing brackets are far apart then the search is probably giving up before a match is found. You can increase the number of lines searched by changing the _maxScanLines_ preference - see the notes above for details. The other likely reason for not finding a match is that there isn't one because (a) it doesn't physically exist in the file or (b) the brackets are mismatched. Consider this example:

```
 (  {     ( )   ) }
```

The first opening bracket is closed too soon so Brackets considers that it does not have a matching closing bracket and no match will be found.
Failure to find a match is a good indication that the logic of your code is broken.
Lastly, if the _maxScanLineLength_ preference is set too low then some lines in your file may not be getting included in the search. See _Preferences_ above for more details.

**Q: How can I change the hotkey your extension uses?**  
A: It can be difficult to choose a hotkey for an extension. While it's easy enough to avoid the keys assigned by Brackets itself, it's impossible to know what other extensions may be using (or even what external apps might be listening for). If you find that the default key combination (CMD-ALT-] / CTRL-ALT-]) is not suitable you can override my choice by editing your _User Key Map_ file. This is done using the _Debug_... _Open User Key Map_ menu item and adding in a new entry like this in the _overrides_ section:

```
"Cmd-Shift-Z": "davidwaterston.goToMatchingBracket.findMatch"
```

where you would replace the _Cmd-Shift-Z_ with your preferred key combination.
You can find more details about the User Key Map file and the available key combinations at [https://github.com/adobe/brackets/wiki/User-Key-Bindings](https://github.com/adobe/brackets/wiki/User-Key-Bindings)

**Q: When will the extension be available in "my language"?**  
A: I'm aware that the appearance of the English text for the menu item I add can be very jarring if you're using Brackets in another language. I'm adding translations whenever I can but I rely on finding a native speaker who can translate the text. If you're able to translate the text for me please submit a pull request or let me know by commenting in [_Issues_](https://github.com/davidwaterston/goto-matching-bracket/issues).

**Q: The translation you have for "my language" is totally wrong!**  
A: That's not really a question but, if you think one of the translations is wrong or could be improved, please let me know by commenting in [_Issues_](https://github.com/davidwaterston/goto-matching-bracket/issues).


## Verifying Releases
I use <a href="http://semver.org" target="_blank" alt="Semantic Versioning">Semantic Versioning</a> to number releases. Each release is tagged with the appropriate version number and signed using <a href="https://www.gnupg.org" target="_blank" alt="Gnu Privacy Guard (GPG)">Gnu Privacy Guard (GPG)</a>. The public key used to sign releases is
```
Name: David Waterston
Email: david@davidwaterston.com
Key ID: A7AD9C85
Signature: 71A9 DC13 447A 1E4F C6EB  5D64 DE08 A991 A7AD 9C85
```
This public key is included in the repository with a SHA1 of 16d013451476fa4a1a67d6ad4b90583e205b53b1.
After cloning the repo, and assuming you have GPG installed correctly, you can import this key into your keychain
```
git cat-file blob pubkey | gpg --import
```
When this public key is successfully imported, you can use it to verify the integrity of any of the tagged releases of this repo
```
git tag -v v1.7.0
```
which should produce output similar to:
```
object 04f37a55784c1f3abc2cf927a935a488aa954035
type commit
tag v1.4.0
tagger David Waterston <david@davidwaterston.com> 1427387056 +0000

Updated the Command Id of the extension to be a more standards-compliant.

Updated the Command Id of the extension to be a more standards-compliant 'davidwaterston.goToMatchingBracket.findMatch' (notice the addition of the 'findMatch'). The main reason for this was to ensure that the extension name would appear when displayed in redmunds Display Shortcuts extension which adds a Show Shortcuts item to the Brackets Help menu.
Also rewrote the text in the README.md for the maxScanLineLength preference as my understanding of this was completely wrong.
gpg: Signature made Thu 26 Mar 16:24:16 2015 GMT using RSA key ID A7AD9C85
gpg: Good signature from "David Waterston <david@davidwaterston.com>" [ultimate]
```
The important thing to notice here is that the RSA key ID matches mine (A7AD9C85) and the line that says that this is a good signature.

The public key can further be verified by checking the details held on <a href="http://pgp.mit.edu/pks/lookup?search=david%40davidwaterston.com&op=index&fingerprint=on&exact=on" target="_blank" alt="pgp.mit.edu">pgp.mit.edu</a> or on my <a href="https://keybase.io/davidwaterston/" target="_blank" alt="pgp.mit.edu">Keybase</a> page.

##Missing a feature?
[Add your idea](http://feathub.com/davidwaterston/goto-matching-bracket/features/new) or [vote on your favorite feature](http://feathub.com/davidwaterston/goto-matching-bracket) to be implemented:

[![Feature Requests](http://feathub.com/davidwaterston/goto-matching-bracket?format=svg)](http://feathub.com/davidwaterston/goto-matching-bracket)

## Future Plans
Current and planned work for this repo is public and detailed in [Waffle](https://waffle.io/davidwaterston/goto-matching-bracket).

## License
Copyright (c) 2015 David Waterston. All rights reserved.
Distributed under an MIT license. See the [LICENSE](https://github.com/davidwaterston/goto-matching-bracket/blob/master/LICENSE) file for more details.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]


## [2.2.1] - 2016-01-13

### Changed
-
Updated the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) to fix the instructions for changing the hotkey used to invoke this extension. This had referenced "davidwaterston.goto-matching-bracket.findMatch" when it should be "davidwaterston.goToMatchingBracket.findMatch".


## [2.2.0] - 2015-10-18
### Added
- This CHANGELOG.md file, formatted in the convention detailed in [Keep a CHANGELOG](http://keepachangelog.com).
- Proposed features/changes can be now be suggested/voted for at [FeatHub](http://feathub.com/davidwaterston/goto-matching-bracket). The [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) now has a new section - [_Missing a feature?_](https://github.com/davidwaterston/goto-matching-bracket#missing-a-feature) - which links directly to this project in [FeatHub](http://feathub.com/davidwaterston/goto-matching-bracket) and includes a dynamic list of the features suggested / votes cast so far.
- Added badge and new text section - [_Future Plans_](https://github.com/davidwaterston/goto-matching-bracket#future-plans) - in [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) that links to this project in [Waffle.io](https://waffle.io/davidwaterston/goto-matching-bracket).

### Changed
- Removed the _enabled_ user preference. As Brackets 1.4+ has the built-in ability to temporarily disable an extension, this feature is no longer required.
- Removed the _Release History_ section of the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) as it is now replaced by this CHANGELOG file.
- Replaced hardcoded version number badge in [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) with one that gets latest version number from GitHub.
- Replaced remaining HTML badges in [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) with Markdown equivalents.
- Cosmetic change to the waffle.io badge in [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md); label is now 'in progress' rather than 'In Progress' to make it more consistent with the other badges
- Added link to my [Keybase](https://keybase.io/davidwaterston) profile from [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) so that my GPG key can more easily be confirmed.


## [2.1.0] - 2015-04-24
### Added
- French translation added, thanks to [@PhiLhoSoft](https://github.com/PhiLhoSoft)

![goto-matching-bracket_v2 1 0](https://cloud.githubusercontent.com/assets/876545/7787797/9c254ae0-0216-11e5-8ea1-ac51d4da5752.png)

### Changed
- The 'charsToMatch' object was tidied up as it could be confusing to anyone reading the code.


## [2.0.1] - 2015-04-23

### Fixed
- Restored a comma that was inadvertantly removed in the previous release, rendering the [package.json](https://github.com/davidwaterston/goto-matching-bracket/blob/master/package.json) file invalid.


## [2.0.0] - 2015-04-23

### Changed
- __Breaking change__: With the release of Brackets 1.3 the original keyboard shortcut used by this extension - Cmd-Alt-] (Mac) or Ctrl-Alt-] (Windows) - has been taken by the Brackets Expand Current command. This release changes the default shortcut to be Cmd-Alt-RightArrow (Mac) or Ctrl-Alt-RightArrow (Windows) and confirms that the extension continues to work in v1.3.
- [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) updated to reflect new hotkeys

### Fixed
- [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) updated to fix the explanation for how to change the hotkey as this was previously incorrect


## [1.7.0] - 2015-04-06

### Added
- Added a new user preference, _enabled_, which allows the extension to be temporarily turned off. This can be useful if you suspect the extension is causing a problem with Brackets or another extension. The default for this setting is _true_ (the extension is enabled).

- Added details of the new _enabled_ user preference to the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md).

### Changed
- Added more detail to the [_Install_](https://github.com/davidwaterston/goto-matching-bracket#install) section of the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md).


## [1.6.0] - 2015-03-31

### Added
- Added an .editorconfig file as specified in [EditorConfig](http://editorconfig.org).
This file helps developers define and maintain consistent coding styles between different editors and IDEs. It has no impact on the functionality of this extension but is intended for use by developers who might work on the source code.

## [1.5.0] - 2015-03-29

### Added
- Added a [License](https://github.com/davidwaterston/goto-matching-bracket/blob/master/LICENSE) file. This replaces a large part of the _License_ section of the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md). The license itself is unchanged.
- Added badges to the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md) to indicate the version number of the current release, the license type and an indicator that the releases are signed using Gnu Privacy Guard (GPG).
- Added a new section to the [README.md](https://github.com/davidwaterston/goto-matching-bracket/blob/master/README.md), [_Verifying Releases_](https://github.com/davidwaterston/goto-matching-bracket#verifying-releases), which details how release tags are signed and how they can be verified to ensure the integrity of the repo.

### Changed

main.js
/*
 * Brackets 'go to matching bracket' extension.
 *
 * Copyright (c) 2015 David Waterston. All rights reserved.
 * Distributed under an MIT license:
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

/*jslint vars: true, plusplus: true, devel: true, regexp: true, nomen: true, indent: 4, maxerr: 50 */
/*global define, brackets */

define(function (require) {

    "use strict";

    var CommandManager = brackets.getModule("command/CommandManager");
    var EditorManager = brackets.getModule("editor/EditorManager");
    var Menus = brackets.getModule("command/Menus");
    var PreferencesManager = brackets.getModule("preferences/PreferencesManager");
    var Strings = require("strings");

    var preferences;
    var preferencesId = "com.github.davidwaterston.gotoMatchingBracket";
    var defaultPreferences = {maxScanLineLength: 10000, maxScanLines: 5000};

    var charsToMatch = {
        "(": true,
        ")": true,
        "[": true,
        "]": true,
        "{": true,
        "}": true
    };


    function loadPreferences() {

        preferences = PreferencesManager.getExtensionPrefs(preferencesId);
        preferences.definePreference("maxScanLineLength", "integer", defaultPreferences.maxScanLineLength);
        preferences.definePreference("maxScanLines", "integer", defaultPreferences.maxScanLines);

    }


    function gotoMatchingBracket() {

        var editor = EditorManager.getActiveEditor();
        var cursorPos = editor.getCursorPos();
        var line = editor.document.getLine(cursorPos.line);
        var validChar = (charsToMatch[line.charAt(cursorPos.ch)]) || (cursorPos.ch >= 0 && charsToMatch[line.charAt(cursorPos.ch - 1)]);

        if (validChar) {
            var matchingBrace = editor._codeMirror.findMatchingBracket(
                editor.getCursorPos(),
                false,
                {
                    maxScanLineLength: preferences.get("maxScanLineLength"),
                    maxScanLines: preferences.get("maxScanLines")
                }
            );
            if (matchingBrace && matchingBrace.match) {
                editor.setCursorPos(matchingBrace.to.line, matchingBrace.to.ch + 1);
            }
        }

    }


    function init() {

        var COMMAND_ID = "davidwaterston.goToMatchingBracket.findMatch";
        CommandManager.register(Strings.MENU_NAVIGATE_GOTO_MATCHING_BRACKET, COMMAND_ID, gotoMatchingBracket);

        var menu = Menus.getMenu(Menus.AppMenuBar.NAVIGATE_MENU);
        menu.addMenuItem(
            COMMAND_ID,
            "Ctrl-Alt-Right",
            Menus.LAST_IN_SECTION,
            Menus.MenuSection.NAVIGATE_GOTO_COMMANDS
        );

    }


    loadPreferences();

    init();

});
